File: php/143-reorder-list.php
<?php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class Solution 
{
    /**
     * @param ListNode $head
     * @return NULL
     */
    public function reorderList($head) 
    {
        if (!$head || !$head->next) {
            return $head;
        }
        $fast = $slow = $prev = $head;
        while ($fast && $fast->next) {
            $prev = $slow;
            $slow = $slow->next;
            $fast = $fast->next->next;
        }
        $l2 = $prev->next;
        $l1 = $head;
        //cut
        $prev->next = null;
        //reverse l2
        $l2 = $this->reverse($l2);
        $newHead = new ListNode(0);
        $newHeadPtr = $newHead;
        while ($l1 && $l2) {
            $newHeadPtr->next = $l1;
            $l1 = $l1->next;
            $newHeadPtr = $newHeadPtr->next;
            $newHeadPtr->next = $l2;
            $l2 = $l2->next;
            $newHeadPtr = $newHeadPtr->next;
        }
        while ($l2) {
            $newHeadPtr->next = $l2;
            $newHeadPtr = $newHeadPtr->next;
            $l2 = $l2->next;
        }
        return $newHeadPtr->next;
    }

    public function reverse($head)
    {
        $node = $head;
        $prev = null;
        while ($node) {
            $next = $node->next;
            $node->next = $prev;
            $prev = $node;
            $node = $next;
        }
        return $prev;
    }
}

File: go/143-reorder-list.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reorderList(head *ListNode) {
	if head == nil || head.Next == nil {
		return
	}
	slow, fast, prev := head, head, head
	//find middle
	for fast != nil && fast.Next != nil {
		prev = slow
		slow = slow.Next
		fast = fast.Next.Next
	}
	l1 := head
	l2 := slow
	prev.Next = nil
	l2 = reverse(slow)
	newHead := &ListNode{}
	node := newHead
	for l1 != nil && l2 != nil {
		node.Next = l1
		l1 = l1.Next
		node = node.Next
		node.Next = l2
		l2 = l2.Next
		node = node.Next
	}
	for l2 != nil {
		node.Next = l2
		node = node.Next
		l2 = l2.Next
	}
	head = newHead.Next
	return
}

func reverse(node *ListNode) *ListNode {
	var prev *ListNode
	for node != nil {
		next := node.Next
		node.Next = prev
		prev = node
		node = next
	}
	return prev
}
